pkg/analysis/maxlength/analyzer.go

```diff
@@ -0,0 +1,225 @@
+package maxlength
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+
+	"github.com/JoelSpeed/kal/pkg/analysis/helpers/markers"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+const (
+	name = "maxlength"
+
+	kubebuilderMaxLength = "kubebuilder:validation:MaxLength"
+	kubebuilderEnum      = "kubebuilder:validation:Enum"
+	kubebuilderFormat    = "kubebuilder:validation:Format"
+
+	kubebuilderItemsMaxLength = "kubebuilder:validation:items:MaxLength"
+	kubebuilderItemsEnum      = "kubebuilder:validation:items:Enum"
+	kubebuilderItemsFormat    = "kubebuilder:validation:items:Format"
+
+	kubebuilderMaxItems = "kubebuilder:validation:MaxItems"
+)
+
+var (
+	errCouldNotGetInspector = errors.New("could not get inspector")
+	errCouldNotGetMarkers   = errors.New("could not get markers")
+)
+
+// Analyzer is the analyzer for the maxlength package.
+// It checks that strings and arrays have maximum lengths and maximum items respectively.
+var Analyzer = &analysis.Analyzer{
+	Name:     name,
+	Doc:      "Checks that all strings formatted fields are marked with a maximum length, and that arrays are marked with max items.",
+	Run:      run,
+	Requires: []*analysis.Analyzer{inspect.Analyzer, markers.Analyzer},
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspect, ok := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	if !ok {
+		return nil, errCouldNotGetInspector
+	}
+
+	markersAccess, ok := pass.ResultOf[markers.Analyzer].(markers.Markers)
+	if !ok {
+		return nil, errCouldNotGetMarkers
+	}
+
+	// Filter to structs so that we can iterate over the fields in a struct.
+	nodeFilter := []ast.Node{
+		(*ast.StructType)(nil),
+	}
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		sTyp, ok := n.(*ast.StructType)
+		if !ok {
+			return
+		}
+
+		if sTyp.Fields == nil {
+			return
+		}
+
+		for _, field := range sTyp.Fields.List {
+			checkField(pass, field, markersAccess)
+		}
+	})
+
+	return nil, nil //nolint:nilnil
+}
+
+func checkField(pass *analysis.Pass, field *ast.Field, markersAccess markers.Markers) {
+	if len(field.Names) == 0 || field.Names[0] == nil {
+		return
+	}
+
+	fieldName := field.Names[0].Name
+	prefix := fmt.Sprintf("field %s", fieldName)
+
+	checkTypeExpr(pass, field.Type, field, nil, markersAccess, prefix, kubebuilderMaxLength, needsStringMaxLength)
```

Comment:
Nit: you could probably reduce the input parameters to `checkTypeExpr` if you instead had it return the information you needed to construct the report information.

Totally personal preference, but having `checkTypeExpr` return the information you need means you only have to change one area where reports are set than multiple if you decided to change it up.

Reply:
At the moment, there are three different error messages that can come out of this function, depending on the route the code takes. One is for `[]string`, one for `[]StringAlias`, and one for the items length itself. At the moment I think refactoring to bring out a set of bools returned (for example), might make this more complex.

I'm not sure this is worth the squeeze right now